lib/season_statistics.rb

```diff
@@ -11,76 +11,56 @@ def initialize
-    @all_games.each do |game_object|
-      if season == game_object["season"]
-        @by_season_game_objects << game_object
+      if game_by_team["game_id"][0..3] == season[0..3]
+        @coach_by_team_id[game_by_team["team_id"]] = game_by_team["head_coach"]
       end
     end
   end
 
-  def total_games_played_by_season(game)
-    @games_played_by_team_id[game["away_team_id"]] += 1
-    @games_played_by_team_id[game["home_team_id"]] += 1
-  end
-
-  def total_wins_for_home(game)
-    @counter_wins_team_id[game["home_team_id"]] += 1
-    @counter_wins_team_id[game["away_team_id"]] += 0
-  end
-
-  def total_wins_for_away(game)
-    @counter_wins_team_id[game["away_team_id"]] += 1
-    @counter_wins_team_id[game["home_team_id"]] += 0
+  def total_games_by_coach_by_season
+    @total_games_by_coach = Hash.new{ |hash, key| hash[key] = 0}
+    @by_season_game_team_objects.each do |game_by_season|
+      @total_games_by_coach[game_by_season["head_coach"]] += 1
+    end
   end
 
-  def total_wins_by_season
-    @by_season_game_objects.each do |season_game_object|
-      total_games_played_by_season(season_game_object)
-      if season_game_object["home_goals"] > season_game_object["away_goals"]
-        total_wins_for_home(season_game_object)
-      elsif season_game_object["away_goals"] > season_game_object["home_goals"]
-        total_wins_for_away(season_game_object)
+  def total_wins_by_coach_by_season
+    @total_wins_by_coach = Hash.new{ |hash, key| hash[key] = 0}
+    @by_season_game_team_objects.each do |game_by_season|
+      if game_by_season["result"] == "WIN"
+        @total_wins_by_coach[game_by_season["head_coach"]] += 1
       end
     end
   end
 
-  def win_percentage_by_season
-    @win_percentage = {}
-    @games_played_by_team_id.each do |id, total_games|
-      @win_percentage[id] = (@counter_wins_team_id[id].to_f / total_games).round(3)
+  def win_percentage_by_coach_by_season
+    @win_percentage_by_coach = {}
+    @total_games_by_coach.each do |coach, total_games|
+      @win_percentage_by_coach[coach] = (@total_wins_by_coach[coach].to_f / total_games).round(3)
     end
   end
 
   def winningest_and_worst_suite(season)
-    create_coach_by_team_id
-    collect_game_objects_by_season(season)
-    total_wins_by_season
-    win_percentage_by_season
+    collect_game_team_objects_by_season(season)
+    create_coach_by_team_id(season)
+    total_games_by_coach_by_season
+    total_wins_by_coach_by_season
+    win_percentage_by_coach_by_season
   end
 
   def winningest_coach(season)
     winningest_and_worst_suite(season)
-    max_team_id = @win_percentage.invert.max[1]
-    winningest_coach = @coach_by_team_id[max_team_id]
+    @win_percentage_by_coach.invert.max[1]
   end
 
   def worst_coach(season)
     winningest_and_worst_suite(season)
-    min_team_id = @win_percentage.invert.min[1]
-    worst_coach = @coach_by_team_id[min_team_id]
+    @win_percentage_by_coach.invert.min[1]
   end
 
```

Comment:
Nice refactoring! Lots of unneeded code removed

Reply:
WOOO!